astroquery/alma/core.py

```diff
@@ -708,22 +708,47 @@ def download_files(self, files, savedir=None, cache=True,
         if savedir is None:
             savedir = self.cache_location
         for fileLink in unique(files):
+            log.debug("Downloading {0} to {1}".format(fileLink, savedir))
             try:
-                log.debug("Downloading {0} to {1}".format(fileLink, savedir))
                 check_filename = self._request('HEAD', fileLink, auth=auth,
                                                stream=True)
                 check_filename.raise_for_status()
-                if 'text/html' in check_filename.headers['Content-Type']:
-                    raise ValueError("Bad query.  This can happen if you "
-                                     "attempt to download proprietary "
-                                     "data when not logged in")
-
-                filename = self._request("GET", fileLink, save=True,
-                                         savedir=savedir,
-                                         timeout=self.TIMEOUT,
-                                         cache=cache,
-                                         auth=auth,
-                                         continuation=continuation)
+            except requests.HTTPError as ex:
+                if ex.response.status_code == 401:
+                    if skip_unauthorized:
+                        log.info("Access denied to {url}.  Skipping to"
+                                 " next file".format(url=fileLink))
+                        continue
+                    else:
+                        raise(ex)
+
+            if 'text/html' in check_filename.headers['Content-Type']:
+                raise ValueError("Bad query.  This can happen if you "
+                                 "attempt to download proprietary "
+                                 "data when not logged in")
+
+            try:
+                filename = re.search("filename=(.*)",
+                                     check_filename.headers['Content-Disposition']).groups()[0]
+            except KeyError:
+                log.info(f"Unable to find filename for {fileLink}  "
+                         "(missing Content-Disposition in header).  "
+                         "Skipping to next file.")
+
+            if savedir is not None:
+                filename = os.path.join(savedir,
```

Comment:
Too bad, I didn't spot this:

`E               UnboundLocalError: local variable 'filename' referenced before assignment`

Reply:
Hmm, the mocks in the tests also need some adjustments. Take away message, I should remote test PRs before merge :)

Reply:
I see.  This is an actual issue I can refer upstream - @andamian if you are able to chime in, it appears that some of the data sets do not have 'filename' metadata in the `Content-Disposition` headers.  Is this an archive bug I should raise via helpdesk?

Reply:
see https://github.com/astropy/astroquery/pull/2098

Reply:
No, it's not the data, but the mock in the test.

Reply:
(adding an easy fix there then messes up the assert that it has to bed triggered twice, and there are more complication with the second failing test)

Reply:
OK, it might not fix the failing test, but this fix _is_ necessary.